subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -4435,8 +4590,12 @@ int lwm2m_engine_context_close(struct lwm2m_ctx *client_ctx)
 
 void lwm2m_engine_context_init(struct lwm2m_ctx *client_ctx)
 {
-	sys_slist_init(&client_ctx->pending_sends);
-	sys_slist_init(&client_ctx->observer);
+	if (client_ctx->conn_status == LWM2M_CLIENT_CONNECTED) {
```

Comment:
This block ensures observer list remains intact when this function is eventually called from `lwm2m_engine_start` upon registration/update failure. It also preserves reinitializing pending list. Both of these lists are also NOT cleared in `lwm2m_engine_context_close` function as part of same goal. Observers must be kept intact so notifications can be generated normally. 
The reason for keeping `pending_sends` intact as well is mentioned in the original post:
> Another case this can happen is for example, reg. update fails before notification timeout. In this case, new registration wipes all in-flight buffers.
To prevent this, we leave the `pending_sends` as is so the notifications which are in-flight continue to be transmitted and timeout eventually. After they timeout, they will be pushed to storage medium. There is one minor problem with this approach and that is order in which these notifications will be stored.